xds/internal/client/xds.go

```diff
@@ -372,22 +426,36 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger)
 // them, and transforms them into a native struct which contains only fields we
 // are interested in.
 func UnmarshalCluster(version string, resources []*anypb.Any, logger *grpclog.PrefixLogger) (map[string]ClusterUpdate, UpdateMetadata, error) {
+	timestamp := time.Now()
```

Comment:
All the `UnmarshalXxx` functions seem to have a very similar structure:
- iterate through the list of `any`, and for each
- make sure the current element is of expected resource type
- unmarshal current element into expected resource type
- actually process the resource, and return error or `XxxUpdate`
- while doing the above, keep accumulating the different errors
- after processing all elements, decide whether to ACK or NACK and create error metadata if required.

Just curious if you thought about defining an interface to be implemented for each resource type and generalizing these functions. Not suggesting that we do that. But just want to know if you tried that, and decided against it.

Reply:
I didn't consider this. It's done now.

We still have the same annoying issue that the return type is different, `map[string]ClusterUpdate` vs `map[string]RouteConfigUpdate`. But I think the end result isn't too bad.

Reply:
A bit easier to review if hiding the whitespace diff.